mlir/lib/Dialect/SCF/IR/SCF.cpp

```diff
@@ -3872,14 +3872,89 @@ struct WhileRemoveDuplicatedResults : public OpRewritePattern<WhileOp> {
     return success();
   }
 };
+
+/// If both ranges contain same values return mappping indices from args1 to
+/// args2. Otherwise return std::nullopt
+static std::optional<SmallVector<unsigned>> getArgsMapping(ValueRange args1,
+                                                           ValueRange args2) {
+  if (args1.size() != args2.size())
+    return std::nullopt;
+
+  SmallVector<unsigned> ret(args1.size());
+  for (auto &&[i, arg1] : llvm::enumerate(args1)) {
+    auto it = llvm::find(args2, arg1);
```

Comment:
I think this will break if there are duplicate values in `args1` and `args2`.

Reply:
Do you have an example where we'd want to find a mapping but there are duplicates?

Reply:
The value range could contain the same value multiple times. eg if the same value is yielded multiple times from the before block. I think then this function no longer returns a permutation. actually, maybe that is fine...

Reply:
One of the args is block args so it will never contain duplicates, the other is `scf.condition` args. There is a pattern to cleanup duplicated `scf.condition` args, but order in which patterns are applied is undetermined. I can add a code to check for duplicated `scf.condition` args and bail out.

Reply:
OR, I can set higher benefit to duplicated cleanup pattern.

Reply:
added duplicates check